packages/loopring_v3/contracts/impl/libexchange/ExchangeWithdrawals.sol

```diff
@@ -37,299 +22,236 @@ library ExchangeWithdrawals
-        address         to,
+        address indexed from,
+        address indexed to,
+        address         token,
         uint96          amount
     );
 
-    function getWithdrawRequest(
+    function forceWithdraw(
         ExchangeData.State storage S,
-        uint index
-        )
-        external
-        view
-        returns (
-            bytes32 accumulatedHash,
-            uint    accumulatedFee,
-            uint32  timestamp
-        )
-    {
-        require(index < S.withdrawalChain.length, "INVALID_INDEX");
-        ExchangeData.Request storage request = S.withdrawalChain[index];
-        accumulatedHash = request.accumulatedHash;
-        accumulatedFee = request.accumulatedFee;
-        timestamp = request.timestamp;
-    }
-
-    function withdraw(
-        ExchangeData.State storage S,
-        uint24  accountID,
+        address owner,
         address token,
-        uint96  amount
+        uint24  accountID
         )
         external
     {
-        require(amount > 0, "ZERO_VALUE");
         require(!S.isInWithdrawalMode(), "INVALID_MODE");
-        require(S.areUserRequestsEnabled(), "USER_REQUEST_SUSPENDED");
-        require(getNumAvailableWithdrawalSlots(S) > 0, "TOO_MANY_REQUESTS_OPEN");
+        require(S.getNumAvailableForcedSlots() > 0, "TOO_MANY_REQUESTS_OPEN");
+        require(accountID != 1 && accountID < ExchangeData.MAX_NUM_ACCOUNTS(), "INVALID_ACCOUNTID");
 
         uint16 tokenID = S.getTokenID(token);
 
+        uint withdrawalFeeETH = S.loopring.forcedWithdrawalFee();
+
         // Check ETH value sent, can be larger than the expected withdraw fee
-        require(msg.value >= S.withdrawalFeeETH, "INSUFFICIENT_FEE");
+        require(msg.value >= withdrawalFeeETH, "INSUFFICIENT_FEE");
 
         // Send surplus of ETH back to the sender
-        uint feeSurplus = msg.value.sub(S.withdrawalFeeETH);
+        uint feeSurplus = msg.value.sub(withdrawalFeeETH);
         if (feeSurplus > 0) {
             msg.sender.sendETHAndVerify(feeSurplus, gasleft());
         }
 
-        // Add the withdraw to the withdraw chain
-        ExchangeData.Request storage prevRequest = S.withdrawalChain[S.withdrawalChain.length - 1];
-        ExchangeData.Request memory request = ExchangeData.Request(
-            sha256(
-                abi.encodePacked(
-                    prevRequest.accumulatedHash,
-                    accountID,
-                    uint16(tokenID),
-                    amount
-                )
-            ),
-            prevRequest.accumulatedFee.add(S.withdrawalFeeETH),
-            uint32(now)
-        );
-        S.withdrawalChain.push(request);
+        require(S.pendingForcedWithdrawals[accountID][tokenID].timestamp == 0, "WITHDRAWAL_ALREADY_PENDING");
 
-        emit WithdrawalRequested(
-            uint32(S.withdrawalChain.length - 1),
-            accountID,
-            tokenID,
-            amount
+        S.pendingForcedWithdrawals[accountID][tokenID].owner = owner;
```

Comment:
what happens if the owner provides an accountID that doesn't belong to him?

Reply:
Withdrawal is processed by the operator (because he is forced to do so), but nothing gets withdrawn.

Reply:
Is it possible that the owner-id combination is correct, but the operator cheats by claiming it's a bad withdrawal (type==3)?

Reply:
The operator can't lie about the account owner for the specified accountID because that data comes straight from the Merkle tree. And the operator also can't lie about who did the forced withdrawal transaction. So we know what we need to know onchain to check if the operator chose the right type.